README.md
# Laika

[![CircleCI](https://circleci.com/gh/MEDIGO/laika.svg?style=shield)](https://circleci.com/gh/MEDIGO/laika) [![Coverage Status](https://coveralls.io/repos/github/MEDIGO/laika/badge.svg)](https://coveralls.io/github/MEDIGO/laika)

Laika is a feature flag/feature toggle service, written in Go, that allows the creation of flags and their activation/deactivation for specific environments. This way it is possible to control in which environments each feature is available. For instance, when a new feature is developed and released, it would make sense if it was only made available, at first, in a testing or Q&A environment, and only later in production. With Laika this can be achieved by simply going to a web page, selecting the feature, and changing its status on the desired environments.

Using Laika in a project thus allows for fast and continuous feature release and deployment.

`laika-php` is a PHP library that handles communication with Laika's API. You can get it [here](https://github.com/MEDIGO/laika-php).

## API Reference

Laika uses CQRS, the query endpoints are as follows.

| Method  | Endpoint                  | Description                |
| ------- | ------------------------- | -------------------------- |
| `GET`   | `/api/health`             | Check the service health   |
| `GET`   | `/api/features`           | List all features          |
| `GET`   | `/api/features/:name`     | Get a feature by name      |
| `GET`   | `/api/environments`       | List all environments      |

The command endpoint is for manipulating data.

| Method  | Endpoint                          | Example body                                                   | Description               |
| ------- | --------------------------------- | -------------------------------------------------------------- | ------------------------- |
| `POST`  | `/api/events/environment_created` | `{"name":"staging"}`                                           | Create a new environment. |
| `POST`  | `/api/events/feature_created`     | `{"name":"feature1"}`                                          | Create a new feature.     |
| `POST`  | `/api/events/user_created`        | `{"username":"admin","password":"secret"}`                     | Create a new user.        |
| `POST`  | `/api/events/feature_toggled`     | `{"feature":"feature1","environment":"staging","status":true}` | Toggle a feature.         |
| `POST`  | `/api/events/feature_deleted`     | `{"name":"feature1"}`                                          | Delete a feature.         |
| `POST`  | `/api/events/environments_ordered`| `{"order":["dev","staging"]}`                                  | Change env display order. |
| `POST`  | `/api/events/environment_deleted` | `{"name":"staging"}`                                           | Delete an environment.    |

## Client

Laika contains a polling HTTP client that allows to easily check for enabled/disabled features on Go code. It can be found in the `client` package. While Laika uses the `vendor` directory to store external dependencies, `client` can be imported without any vendoring.

### Install

```
$ go get -u github.com/MEDIGO/laika/client
```

### Usage

```go
package main

import (
	"log"

	"github.com/MEDIGO/laika/client"
)

func main() {
	c, err := client.NewClient(client.Config{
		Addr:        "127.0.0.1:8000",
		Username:    "my-username",
		Password:    "my-password",
		Environment: "prod",
	})

	if err != nil {
		log.Fatal(err)
	}

	if c.IsEnabled("my-awesome-feature", false) {
		log.Print("IT'S ALIVE!")
	} else {
		log.Print("Move along. Nothing to see here.")
	}
}
```

## Developing

To develop Laika you need to have the following tools installed in your machine:

* [Go](https://golang.org/doc/install)
* [Node.JS](https://nodejs.org/en/download/)
* [Glide](https://github.com/Masterminds/glide)
* [MySQL](https://dev.mysql.com/downloads/installer/)

Then install all the Go and Javascript dependencies with:

```sh
$ make install
```


Build continuously the server and UI with:

```sh
$ make develop
```

And start hacking!

```sh
$ open http://localhost:8000
```

## Testing

The whole test suite can be executed with:

```
$ make test
```

Some test require a MySQL instance, you can pass the configuration to them with the following
environment variables:

```
LAIKA_TEST_MYSQL_HOST=localhost
LAIKA_TEST_MYSQL_PORT=3306
LAIKA_TEST_MYSQL_USERNAME=root
LAIKA_TEST_MYSQL_PASSWORD=root
LAIKA_TEST_MYSQL_DBNAME=test
```

## Current state of the project

In the current release of Laika, it is possible to create feature flags and enable/disable them in the existing environments.

### Wishlist

- Specify country access (e.g. feature only enabled in Germany).
- Specify user access with percentage (e.g. feature only enabled for 30% of the user base).
- Have a field for environment creation on the web page.
- History for flag status changes.

## Copyright and license

Copyright © 2016 MEDIGO GmbH.

Laika is licensed under the MIT License. See [LICENSE](LICENSE) for the full license text.


CHANGES.md
# Changes

## ?.?.?

### New features

* Environments can be deleted through the API.

### Bugfixes and minor changes

* Slack notifier is now less spammy.
* Fix unstable UI sort edge case with old data.


## 1.0.0

Major changes on frontend and backend. The API contains breaking changes, but neither the Go nor the PHP client are affected by these changes.

### Major changes

* Switched frontend to React.
* Switched backend to an event-sourcing architecture (migration is automatic).
* API no longer returns numeric IDs, but uses the name of environments and features for identity.

### New features

* Features can be deleted from the UI.
* Always migrate database on startup.
* Increased creation timestamp precision from days to seconds.
* Order environments by creation time in the UI.
* Add API endpoint to order environments manually.
* Auto focus input fields in the UI.

### Bugfixes and minor changes

* Unknown API routes now properly return 404.
* Unknown frontend routes now redirect to the home page.


## 0.8.0

This is the first versioned release.

### New features

* Features are now sorted in reverse chronological order.

### Bugfixes and minor changes

* Names with spaces, slashes, and various other characters no longer break the UI.
* Fix typescript error with moment.js.


## 0.0.0

This corresponds to everything up to commit `3637a6e` (inclusive) made on Jan 29, 2018.


main.go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/DataDog/datadog-go/statsd"
	"github.com/MEDIGO/laika/api"
	"github.com/MEDIGO/laika/notifier"
	"github.com/MEDIGO/laika/store"
	log "github.com/Sirupsen/logrus"
	"github.com/urfave/cli"
	graceful "gopkg.in/tylerb/graceful.v1"
)

func init() {
	log.SetFormatter(&log.JSONFormatter{})
}

func main() {
	app := cli.NewApp()
	app.Author = "MEDIGO GmbH"
	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:   "port",
			Value:  "8000",
			Usage:  "Service port",
			EnvVar: "LAIKA_PORT",
		},
		cli.IntFlag{
			Name:   "timeout",
			Value:  10,
			Usage:  "Shutdown timeout",
			EnvVar: "LAIKA_TIMEOUT",
		},
		cli.StringFlag{
			Name:   "mysql-host",
			Value:  "localhost",
			Usage:  "MySQL host",
			EnvVar: "LAIKA_MYSQL_HOST",
		},
		cli.StringFlag{
			Name:   "mysql-port",
			Value:  "3306",
			Usage:  "MySQL port",
			EnvVar: "LAIKA_MYSQL_PORT",
		},
		cli.StringFlag{
			Name:   "mysql-username",
			Value:  "root",
			Usage:  "MySQL username",
			EnvVar: "LAIKA_MYSQL_USERNAME",
		},
		cli.StringFlag{
			Name:   "mysql-password",
			Value:  "root",
			Usage:  "MySQL password",
			EnvVar: "LAIKA_MYSQL_PASSWORD",
		},
		cli.StringFlag{
			Name:   "mysql-dbname",
			Value:  "laika",
			Usage:  "MySQL dbname",
			EnvVar: "LAIKA_MYSQL_DBNAME",
		},
		cli.StringFlag{
			Name:   "statsd-host",
			Value:  "localhost",
			Usage:  "Statsd host",
			EnvVar: "LAIKA_STATSD_HOST",
		},
		cli.StringFlag{
			Name:   "statsd-port",
			Value:  "8125",
			Usage:  "Statsd port",
			EnvVar: "LAIKA_STATSD_PORT",
		},
		cli.StringFlag{
			Name:   "root-username",
			Usage:  "Root username",
			Value:  "root",
			EnvVar: "LAIKA_ROOT_USERNAME",
		},
		cli.StringFlag{
			Name:   "root-password",
			Usage:  "Root password",
			Value:  "root",
			EnvVar: "LAIKA_ROOT_PASSWORD",
		},
		cli.StringFlag{
			Name:   "slack-webhook-url",
			Usage:  "Slack webhook URL",
			EnvVar: "LAIKA_SLACK_WEBHOOK_URL",
		},
	}
	app.Commands = []cli.Command{
		{
			Name:  "run",
			Usage: "Runs laika's feature flag service",
			Action: func(c *cli.Context) error {
				store, err := store.NewMySQLStore(
					c.GlobalString("mysql-username"),
					c.GlobalString("mysql-password"),
					c.GlobalString("mysql-host"),
					c.GlobalString("mysql-port"),
					c.GlobalString("mysql-dbname"),
				)

				if err != nil {
					return fmt.Errorf("failed to create store: %s", err)
				}

				if err := store.Ping(); err != nil {
					return fmt.Errorf("could not ping database: %s", err)
				}

				if err := store.Migrate(); err != nil {
					return fmt.Errorf("failed to migrate store schema: %s", err)
				}

				if _, err := store.State(); err != nil {
					return fmt.Errorf("failed to compute initial state: %s", err)
				}

				stats, err := statsd.New(c.GlobalString("statsd-host") + ":" + c.GlobalString("statsd-port"))
				if err != nil {
					return fmt.Errorf("failed to create Statsd client: %s", err)
				}

				notifier := notifier.NewSlackNotifier(c.GlobalString("slack-webhook-url"))

				server, err := api.NewServer(api.ServerConfig{
					RootUsername: c.GlobalString("root-username"),
					RootPassword: c.GlobalString("root-password"),
					Store:        store,
					Stats:        stats,
					Notifier:     notifier,
				})
				if err != nil {
					return fmt.Errorf("failed to create server: %s", err)

				}

				log.Info("Starting server on port ", c.GlobalString("port"))
				graceful.Run(":"+c.GlobalString("port"), time.Duration(c.Int("timeout"))*time.Second, server)
				log.Info("Server exiting")
				return nil
			},
		},
	}

	if err := app.Run(os.Args); err != nil {
		log.Fatal(err)
	}
}
